server/Arcadia.Assistant.Notifications.Push/PushNotificationsActor.cs

```diff
@@ -52,13 +52,44 @@ private Task SendPushNotification(PushNotification message)
         {
             this.logger.Debug("Push notification message received");
 
-            var jsonMessage = this.SerializeNotification(message);
-            this.logger.Debug($"Serialized notification message: {jsonMessage}");
+            if (!message.DevicePushTokens.Any())
+            {
+                this.logger.Debug("Push notification message doesn't contain target devices and won't be sent");
+                return Task.CompletedTask;
+            }
+
+            var androidPushTask = this.SendApplicationPushNotification(message, PushDeviceTypes.Android);
+            var iosPushTask = this.SendApplicationPushNotification(message, PushDeviceTypes.Ios);
+
+            return Task.WhenAll(androidPushTask, iosPushTask);
+        }
+
+        private Task SendApplicationPushNotification(PushNotification message, string deviceType)
+        {
+            var pushNotificationPayload = new PushNotificationPayload
+            {
+                Content = message.Content,
+                Target = new PushNotificationTarget
+                {
+                    DevicePushTokens = message.DevicePushTokens
+                        .Where(x => x.DeviceType == deviceType)
+                        .Select(x => x.Token)
+                        .ToList()
+                }
+            };
+
+            if (!pushNotificationPayload.Target.DevicePushTokens.Any())
+            {
+                return Task.CompletedTask;
+            }
 
-            var requests = this.pushSettings.ApplicationPushUrls
-                .Select(url => this.SendPushNotificationRequest(url, jsonMessage));
+            var jsonMessage = this.SerializeNotification(pushNotificationPayload);
+            this.logger.Debug($"Serialized {deviceType} notification message: {jsonMessage}");
 
-            return Task.WhenAll(requests);
+            var pushUrl = deviceType == PushDeviceTypes.Android
```

Comment:
pass a push url from calling code so you won't need a check?

Reply:
No, `deviceType` is used also to filter target device tokens to not send Android notification to iOS devices and vice versa.

Reply:
Nevermind, fixed.